### PR TITLE
Scaffold a basic app for node & fetch it via npm.

### PR DESCRIPTION
**Pull request by ngryman**
Related to #4.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,3 @@
 node_modules
-app/templates
+app/templates/apache
```
Comment (alrra) [inline .gitignore]: @ngryman this should be left to `app/templates`

Comment (ngryman) [inline .gitignore]: I'm using `templates` for copying `package.json` and `index.js` to the project.
Where should I put those files?

Comment (ngryman) [inline .gitignore]: ping @alrra.

File: app/index.js [hunk 2]
```diff
@@ -3,6 +3,7 @@
 var fs = require('fs');
 var path = require('path');
 var util = require('util');
+var npm = require("npm");
```
Comment (sindresorhus) [inline app/index.js]: npm is a HUGE dependency. better to use the yeoman API method `installDependencies()`.

Comment (arthurvr) [inline app/index.js]: Btw, avoid double quotes.

File: app/templates/index.js [hunk 3]
```diff
@@ -0,0 +1,4 @@
+var h5bp = require('h5bp');
+
+var app = h5bp.createServer({ root: __dirname + '/public' });
+app.listen(3000);
```
Comment (arthurvr) [inline app/templates/index.js]: Missing newline

Comment (ngryman) [inline app/templates/index.js]: Can you elaborate?

Comment (arthurvr) [inline app/templates/index.js]: @ngryman Our [`.editorconfig`](https://github.com/h5bp/generator-server-configs/blob/master/.editorconfig#L8) states we want newlines at the end of a file. This file doesn't have one. 
scaffold_a_basic_app_for_node___fetch_it_via_npm__by_ngryman_ _pull_request__5_ _h5bp_generator-server-configs

File: app/templates/package.json [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+	"author": {
+		"name": "",
+		"email": "",
+		"url": ""
+	},
+	"main": "index.js",
+	"engines": {
+		"node": ">=0.8.0"
+	}
+}
```
Comment (sindresorhus) [inline app/templates/package.json]: use this package.json instead https://github.com/sindresorhus/node-module-boilerplate/blob/master/boilerplate/package.json

File: app/index.js [hunk 5]
```diff
@@ -80,13 +80,19 @@ Generator.prototype.copyFiles = function (arg) {
 
     var fetch = function (configs) {
 
-        var configsPath = path.join(this.sourceRoot(), configs.server);
+        // node server config is fetched via npm
+        if ( configs.server == 'node' ) {
```
Comment (sindresorhus) [inline app/index.js]: no space here

File: app/index.js [hunk 6]
```diff
@@ -159,4 +165,13 @@ Generator.prototype.copyFiles = function (arg) {
 
 };
 
+Generator.prototype.installNodeConfig = function() {
```
Comment (sindresorhus) [inline app/index.js]: `function ()`

File: app/index.js [hunk 7]
```diff
@@ -159,4 +165,13 @@ Generator.prototype.copyFiles = function (arg) {
...
+  if (!this.isNodeConfig) return;
```
Comment (sindresorhus) [inline app/index.js]:
``` js
if (!this.isNodeConfig) {
  return;
}
```

File: app/index.js [hunk 8]
```diff
@@ -80,13 +80,19 @@ Generator.prototype.copyFiles = function (arg) {
 
     var fetch = function (configs) {
 
-        var configsPath = path.join(this.sourceRoot(), configs.server);
+        // node server config is fetched via npm
+        if ( configs.server == 'node' ) {
+            this.copy('package.json', 'package.json');
+            this.copy('index.js', 'index.js');
```
Comment (sindresorhus) [inline app/index.js]: one argument is enough if the src/dest has the same name

**Comment by sindresorhus** [comment]
@ngryman Can you fix the merge conflict and address the inline comments so we can get this merged?

**Comment by ngryman** [comment]
@sindresorhus Yes sorry I've been ultra busy, will do :)

**Comment by arthurvr** [comment]
ping @ngryman

**Comment by ngryman** [comment]
Hi,

After a long long time ;)
- fixes based on your feedback.
- updated to latest generator code.
- used your `package.json` template @sindresorhus. Some tweaks:
  - removed `<% ... %>` from it as it crashes `npm` when installing.
  - removed `files` entry because inadvertent users may forget to add other project's files.
  - removed `mocha` to let them choose what they want.
- increased tests timeout to let `node-sass` compile when installing `server-configs-node`.

But, tests are failing on `iojs` because `node-sass` dependency is too old. I'm about to make a new release of `server-configs-node` with bumped dependencies. So I'll ping you when it's ready, tests will pass. I'll also squash everything when it's ok.

File: app/templates/package.json [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+		"start": "node index.js"
+  },
+  "keywords": [
+    ""
+  ],
+  "dependencies": {
+    "h5bp": "^0.1.2"
+  },
+  "devDependencies": {}
+}
```
Comment (sindresorhus) [inline app/templates/package.json]: Since this is for a server, we can drop everything except for `scripts` and `dependencies`.

Comment (ngryman) [inline app/templates/package.json]: @sindresorhus Like everything everything? (name, version, ...)

Comment (sindresorhus) [inline app/templates/package.json]: Yup

File: app/templates/package.json [hunk 10]
```diff
@@ -0,0 +1,8 @@
+{
+  "scripts": {
+    "start": "node index.js"
+  },
+  "dependencies": {
+    "h5bp": "^0.1.3"
+  }
+}
```
Comment (arthurvr) [inline app/templates/package.json]: Let's don't call this file `package.json` as NPM will try to use it and fail. Just call it `_package.json` and copy it to `package.json`.

File: app/templates/index.js [hunk 11]
```diff
@@ -0,0 +1,4 @@
+var h5bp = require('h5bp');
```
Comment (arthurvr) [inline app/templates/index.js:3]: Let's `use strict`

File: app/index.js [hunk 12]
```diff
@@ -118,17 +118,26 @@ var Generator = module.exports = yeoman.generators.Base.extend({
 		var configsPath = path.join(this.sourceRoot(), configs.server);
 		var destPath = this.destinationRoot();
 
+		// node server config is fetched via npm
+		if (configs.server == 'node') {
+			this.copy('package.json');
+			this.copy('index.js');
+			done();
+
+			this.npmInstall(['h5bp'], {
```
Comment (arthurvr) [inline app/index.js]: I wonder if we shouldn't use `installDependencies()` here. Also let's support a `--skip-install` option, nearly every yeoman generator has that.

File: app/index.js [hunk 13]
```diff
@@ -118,17 +118,26 @@ var Generator = module.exports = yeoman.generators.Base.extend({
 		var configsPath = path.join(this.sourceRoot(), configs.server);
 		var destPath = this.destinationRoot();
 
+		// node server config is fetched via npm
+		if (configs.server == 'node') {
```
Comment (arthurvr) [inline app/index.js]: triple equal

File: .gitignore [hunk 14]
```diff
@@ -1,3 +1,5 @@
 node_modules
 app/templates
```
Comment (arthurvr) [inline .gitignore]: Just wondering but why do we even have this rule?

File: app/templates/index.js [hunk 15]
```diff
@@ -0,0 +1,4 @@
...
+
+var app = h5bp.createServer({ root: __dirname + '/public' });
```
Comment (arthurvr) [inline app/templates/index.js]: There isn't really a thing as a h5bp style guide so I think we should use the code style @sindresorhus prefers. Which is, if I'm correct, `h5bp.createServer({root: __dirname + '/public'});` (the spacing)

File: package.json [hunk 16]
```diff
@@ -18,7 +18,7 @@
   "main": "app/index.js",
   "repository": "h5bp/generator-server-configs",
   "scripts": {
-    "test": "mocha --timeout 50000"
+    "test": "mocha --timeout 60000"
```
Comment (arthurvr) [inline package.json]: Why is this needed? That seems quite a bunch of time.

Comment (ngryman) [inline package.json]: > increased tests timeout to let node-sass compile when installing server-configs-node.

Compilation may take some time, on my machine (1 yo MBP), it took more that 50sec.

Comment (arthurvr) [inline package.json]: Yeah but generally with testing Yeoman generators you don't install the deps, and just use `skip-install` in the tests.

Comment (ngryman) [inline package.json]: Then I don't see the point of this PR ;)
Anyway, we are speaking about +10sec timeout on tests, is this really important?

Comment (arthurvr) [inline package.json]: No, it isn't too important (though it's 10 seconds of productivity gone).

Installing the dependencies should happen when the users uses the generator, though, why is installing the dependencies really needed to run the tests? Generally we don't do that at any yeoman generators.

Comment (ngryman) [inline package.json]: As installing `h5bp` dep is one of the point of this PR, I thought that was a good idea to test if it was successful. If you guys think it's overkill, I can remove it by skipping install in the tests.

Comment (arthurvr) [inline package.json]: cc @sindresorhus for his thoughts. I personally don't think it adds that much value. I do consider tests running for really long a problem though. 10 seconds extra is quite a bunch.

Comment (ngryman) [inline package.json]: PS: testing it revealed that it doesn't work on `iojs`, which is a good thing to know.

Comment (arthurvr) [inline package.json]: @ngryman that's something that should have been tested at the server-configs repo instead :)

Comment (ngryman) [inline package.json]: Yes sure, but it's good to know that a dependency you have doesn't work so you can file a bug and tell your users why it doesn't work ;)

Comment (sindresorhus) [inline package.json]: I agree with @arthurvr. Not the responsibility of this module to test that.

**Comment by sindresorhus** [comment]
Tests are failing.

**Comment by sindresorhus** [comment]
And why is does `h5bp` have so many dependencies?

`webmake, uglify-js, express, stylus, requirejs, node-sass, less, async, clean-css`

Feels like it's doing way too much.

**Comment by arthurvr** [comment]
> Feels like it's doing way too much.

Yup. A magic server doing all the things is pretty much what server-configs-node has always been :( We should refresh it, one day, ..
